Sources/SwiftParser/Attributes.swift

```diff
@@ -166,8 +166,34 @@ extension Parser {
   }
 
   mutating func parseAttribute(argumentMode: AttributeArgumentMode, parseArguments: (inout Parser) -> RawAttributeSyntax.Argument) -> RawAttributeListSyntax.Element {
-    let (unexpectedBeforeAtSign, atSign) = self.expect(.atSign)
-    let attributeName = self.parseType()
+    var (unexpectedBeforeAtSign, atSign) = self.expect(.atSign)
+    var attributeName = self.parseType()
+    var unexpectedBetweenAtSignAndAttributeName: RawUnexpectedNodesSyntax?
+
+    if atSign.trailingTriviaByteLength != 0 {
+      unexpectedBeforeAtSign = RawUnexpectedNodesSyntax(
+        combining: unexpectedBeforeAtSign,
+        atSign,
+        arena: self.arena
+      )
+      atSign = RawTokenSyntax(
+        kind: .atSign,
+        text: atSign.tokenText,
+        leadingTriviaPieces: atSign.leadingTriviaPieces,
+        presence: .missing,
+        arena: self.arena
+      )
+    } else if attributeName.raw.leadingTriviaByteLength != 0 {
+      unexpectedBetweenAtSignAndAttributeName = RawUnexpectedNodesSyntax(attributeName)
+      // `withLeadingTrivia` only returns `nil` if there is no token in `attributeName`.
+      // But since `attributeName` has leadingTriviaLength != 0 there must be trivia and thus a token.
+      // So we can safely force-unwrap here.
+      attributeName = attributeName
+        .raw
+        .withLeadingTrivia([], arena: self.arena)!
+        .as(RawTypeSyntax.self)!
```

Comment:
I suspect this is where the round-trip failures is coming from. You have already stored `attributeName` in `unexpectedBetweenAtSignAndAttributeName` and now you’re adding it again with `SourcePresence.present`. Although it doesn’t quite match the failure message, we have the node twice in the tree now, once with trivia and once without trivia.

The best solution here is a little bit ugly but it’s the best I can come up with: We want to shove the first token of the `attributeName` (which contains the leading trivia) into `unexpectedBetweenAtSignAndAttributeName` and instead synthesize a new missing token that doesn’t have leading trivia for the first token in `attributeName`. 

Getting the first token to put into `unexpectedBetweenAtSignAndAttributeName` should be as easy as doing `attributeName.firstToken(viewMode: .sourceAccurate)`.

To replace the first token in `attributeName` you need to write a `SyntaxRewriter`. On the first token that it encounters, it should generate a corresponding missing token without leading trivia. After that, it doesn’t rewrite any further tokens. `SyntaxRewriter` operates on the `Syntax`, not `RawSyntax` level, so you will need to convert `attributeName` to a `Syntax` and extract the `RawSyntax` from the rewritten node. Also, you need to make sure that the rewritten first token is allocated in the same `SyntaxArena` as all the other tokens. https://github.com/apple/swift-syntax/pull/1383 contains a change to `SyntaxRewriter` that allows that.

Reply:
> Getting the first token to put into unexpectedBetweenAtSignAndAttributeName should be as easy as doing attributeName.firstToken(viewMode: .sourceAccurate).

The only way I managed to do this was to first declare `RawSyntax.firstToken` as `@_spi(RawSyntax) public` and then doing the following:
```swift
    } else if attributeName.raw.leadingTriviaByteLength != 0,
      let firstAttributeNameToken = attributeName.raw.firstToken(viewMode: .sourceAccurate)
    {
      unexpectedBetweenAtSignAndAttributeName = RawUnexpectedNodesSyntax(
        firstAttributeNameToken.withPresence(.present, arena: self.arena)
      )
      ...
```

Reply:
Can you push your changes and then I’ll take a look at them.